core/src/upgrade/from_fn.rs

```diff
@@ -0,0 +1,111 @@
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::upgrade::{InboundUpgrade, OutboundUpgrade, ProtocolName, UpgradeInfo};
+use crate::{Endpoint, Negotiated};
+
+use futures::prelude::*;
+use std::{iter, pin::Pin};
+
+/// Initializes a new [`FromFnUpgrade`].
+///
+/// # Example
+///
+/// ```
+/// # use libp2p_core::transport::{Transport, MemoryTransport};
+/// # use libp2p_core::upgrade;
+/// # use std::io;
+/// let _transport = MemoryTransport::default()
+///     .and_then(move |out, cp| {
+///         upgrade::apply(out, upgrade::from_fn("/foo/1", move |mut sock, endpoint| async move {
+///             if endpoint.is_dialer() {
+///                 upgrade::write_one(&mut sock, "some handshake data").await?;
+///             } else {
+///                 let handshake_data = upgrade::read_one(&mut sock, 1024).await?;
+///                 if handshake_data != b"some handshake data" {
+///                     return Err(upgrade::ReadOneError::from(io::Error::from(io::ErrorKind::Other)));
+///                 }
+///             }
+///
+///             Ok(sock)
+///         }), cp, upgrade::Version::V1)
+///     });
+/// ```
+///
+pub fn from_fn<P, F, C, Fut, Out, Err>(protocol_name: P, fun: F) -> FromFnUpgrade<P, F>
+where
+    // Note: these bounds are there in order to help the compiler infer types
+    P: ProtocolName + Clone,
+    F: FnOnce(C, Endpoint) -> Fut,
+    Fut: Future<Output = Result<Out, Err>>,
+{
+    FromFnUpgrade { protocol_name, fun }
+}
+
+/// Implements the `UpgradeInfo`, `InboundUpgrade` and `OutboundUpgrade` traits.
+///
+/// The upgrade consists in calling the function passed when creating this struct.
+#[derive(Debug, Clone)]
+pub struct FromFnUpgrade<P, F> {
+    protocol_name: P,
+    fun: F,
+}
+
+impl<P, F> UpgradeInfo for FromFnUpgrade<P, F>
+where
+    P: ProtocolName + Clone,
+{
+    type Info = P;
+    type InfoIter = iter::Once<P>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(self.protocol_name.clone())
+    }
+}
+
+impl<C, P, F, Fut, Err, Out> InboundUpgrade<C> for FromFnUpgrade<P, F>
+where
+    P: ProtocolName + Clone,
+    F: FnOnce(Negotiated<C>, Endpoint) -> Fut,
+    Fut: Future<Output = Result<Out, Err>>,
+{
+    type Output = Out;
+    type Error = Err;
+    type Future = Pin<Box<Fut>>; // TODO: don't box the future
```

Comment:
Do you want to file an issue for the TODOs in here?

Reply:
We could also do #1384 first, I'm not sure.

Reply:
Will rebase on top of #1384 once it's merged.